src/util/std_expr.h

```diff
@@ -4283,47 +4283,81 @@ class infinity_exprt : public nullary_exprt
 };
 
 /// \brief A let expression
-class let_exprt : public ternary_exprt
+class let_exprt : public binary_exprt
 {
 public:
   let_exprt(symbol_exprt symbol, exprt value, const exprt &where)
-    : ternary_exprt(
+    : binary_exprt(
+        multi_ary_exprt(
+          ID_tuple,
+          {std::move(symbol), std::move(value)},
+          typet(ID_tuple)),
+        ID_let,
+        where,
+        where.type())
+  {
+  }
+
+  using symbol_value_sequencet = std::vector<exprt>;
```

Comment:
I know this isn't the first time/only place we are doing this, but we really shouldn't: this should be `exprt::operandst`, unless we are very intentionally (but still somewhat questionably) using a subtype that we know has the same memory layout (as was the case for `binding_exprt`).

Reply:
I don't see much of an alternative to strengthen typing while still exposing something that can be iterated over.
An option could be to invent a custom range-like setup, which would essentially wrap the underlying iterator. That would eliminate the scary casts at the level of the container.

Reply:
I'd be happy to look into this, but given that this is a very global problem, I would declare that to be outside of the scope of this PR.

Reply:
Well, can we at least not make it any worse? I.e., not introduce yet another local definition of what we believe to be the type of `exprt::operands()`?

Reply:
Without that, it's not possible to type the symbols.